File: server/src/main/java/org/jetbrains/bsp/bazel/server/ServerInitializer.java
package org.jetbrains.bsp.bazel.server;

import io.grpc.Server;
import io.vavr.collection.Iterator;
import java.io.PrintWriter;
import java.nio.file.Files;
import java.nio.file.Paths;
import java.nio.file.StandardOpenOption;
import java.util.Arrays;
import java.util.concurrent.Executors;
import org.jetbrains.bsp.bazel.commons.Constants;
import org.jetbrains.bsp.bazel.server.bsp.BspIntegrationData;
import org.jetbrains.bsp.bazel.server.bsp.config.BazelBspServerConfig;
import org.jetbrains.bsp.bazel.server.bsp.info.BspInfo;

public class ServerInitializer {

  public static void main(String[] args) {
    if (args.length > 1) {
      System.err.printf(
          "Expected optional path to project view file; got too many args: %s%n",
          Arrays.toString(args));
      System.exit(1);
    }

    var hasErrors = false;
    var stdout = System.out;
    var stdin = System.in;
    var executor = Executors.newCachedThreadPool();

    try {
      var bspInfo = new BspInfo();
      var rootDir = bspInfo.bazelBspDir();
      Files.createDirectories(rootDir);

      var traceFile = rootDir.resolve(Constants.BAZELBSP_TRACE_JSON_FILE_NAME);
      var traceWriter =
          new PrintWriter(
              Files.newOutputStream(
                  traceFile, StandardOpenOption.CREATE, StandardOpenOption.TRUNCATE_EXISTING));

      var bazelBspServerConfig = getBazelBspServerConfig(bspInfo, args);

      var bspIntegrationData = new BspIntegrationData(stdout, stdin, executor, traceWriter);
      var bspServer = new BazelBspServer(bazelBspServerConfig);
      bspServer.startServer(bspIntegrationData);

      Server server = bspIntegrationData.getServer().start();
      bspServer.setBesBackendPort(server.getPort());

      bspIntegrationData.getLauncher().startListening();
      server.awaitTermination();
    } catch (Exception e) {
      e.printStackTrace();
      hasErrors = true;
    } finally {
      executor.shutdown();
    }

    if (hasErrors) {
      System.exit(1);
    }
  }

  private static BazelBspServerConfig getBazelBspServerConfig(BspInfo bspInfo, String[] args) {
    var projectViewPath = Iterator.of(args).headOption().map(Paths::get);
    return new BazelBspServerConfig(projectViewPath, bspInfo);
  }
}

File: server/src/main/kotlin/org/jetbrains/bsp/bazel/server/ServerInitializer.kt
package org.jetbrains.bsp.bazel.server

import org.apache.logging.log4j.LogManager
import org.jetbrains.bsp.bazel.commons.Constants
import org.jetbrains.bsp.bazel.server.benchmark.TelemetryConfig
import org.jetbrains.bsp.bazel.server.bsp.BspIntegrationData
import org.jetbrains.bsp.bazel.server.bsp.info.BspInfo
import org.jetbrains.bsp.bazel.workspacecontext.DefaultWorkspaceContextProvider
import java.io.PrintWriter
import java.nio.file.Files
import java.nio.file.Path
import java.nio.file.StandardOpenOption
import java.util.concurrent.Executors
import kotlin.io.path.Path
import kotlin.system.exitProcess

data class CliArgs(
    val bazelWorkspaceRoot: String,
    val projectViewPath: String,
    val produceTraceLog: Boolean,
)

private val log = LogManager.getLogger(ServerInitializer::class.java)

object ServerInitializer {
    @JvmStatic
    fun main(args: Array<String>) {
        log.info("Starting server with args: ${args.toList()}")

        Runtime.getRuntime().addShutdownHook(
                Thread {
                    ProcessHandle.allProcesses()
                            .filter { it.parent().orElse(null)?.pid() == ProcessHandle.current().pid() }
                            .forEach { it.destroy() }
                }
        )

        val cliArgs = if (args.size != 3) {
            log.error("Wrong number of args, exiting with exit code 1")
            System.err.println("Usage: <bazel workspace root> <project view path> <produce trace log flag>")
            exitProcess(1)
        } else {
            CliArgs(
                bazelWorkspaceRoot = args.elementAt(0),
                projectViewPath = args.elementAt(1),
                produceTraceLog = args.elementAt(2).toBoolean(),
            )
        }
        var hasErrors = false
        val stdout = System.out
        val stdin = System.`in`
        val executor = Executors.newCachedThreadPool()
        try {
            log.info("Initializing server")
            val bspInfo = BspInfo()
            val rootDir = bspInfo.bazelBspDir()
            Files.createDirectories(rootDir)
            val traceFile = rootDir.resolve(Constants.BAZELBSP_TRACE_JSON_FILE_NAME)
            val workspaceContextProvider = DefaultWorkspaceContextProvider(
                workspaceRoot = Path(cliArgs.bazelWorkspaceRoot),
                projectViewPath = Path(cliArgs.projectViewPath)
            )
            val bspIntegrationData = BspIntegrationData(
                stdout,
                stdin,
                executor,
                createTraceWriterOrNull(traceFile, cliArgs.produceTraceLog)
            )
            val bspServer = BazelBspServer(bspInfo, workspaceContextProvider, Path(cliArgs.bazelWorkspaceRoot), TelemetryConfig())
            val launcher = bspServer.buildServer(bspIntegrationData)
            launcher.startListening().get()
            log.info("Server has been initialized")
        } catch (e: Exception) {
            log.error("Server initialization failed", e)
            e.printStackTrace(System.err)
            hasErrors = true
        } finally {
            executor.shutdown()
        }
        if (hasErrors) {
            exitProcess(1)
        }
    }

    private fun createTraceWriterOrNull(
        traceFile: Path,
        createTraceFile: Boolean,
    ): PrintWriter? =
        if (createTraceFile) {
            PrintWriter(
                Files.newOutputStream(
                    traceFile,
                    StandardOpenOption.CREATE,
                    StandardOpenOption.TRUNCATE_EXISTING
                )
            )
        } else null
}
